src/Dhall/Map.hs

```diff
@@ -0,0 +1,312 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+-- | `Map` type used to represent records and unions
+
+module Dhall.Map
+    ( -- * Type
+      Map
+
+      -- * Construction
+    , empty
+    , singleton
+    , fromList
+
+      -- * Sorting
+    , sort
+    , isSorted
+
+      -- * Insertion
+    , insert
+    , insertWith
+
+      -- * Deletion/Update
+    , delete
+
+      -- * Query
+    , lookup
+    , member
+    , uncons
+
+      -- * Combine
+    , union
+    , unionWith
+    , intersection
+    , intersectionWith
+
+      -- * Traversals
+    , mapWithKey
+    , traverseWithKey
+    , traverseWithKey_
+    , foldMapWithKey
+
+      -- * Conversions
+    , toList
+    , toMap
+    , keys
+    ) where
+
+import Control.Applicative ((<|>))
+import Data.Data (Data)
+import Prelude hiding (lookup)
+
+import qualified Data.Map
+import qualified Data.Set
+
+{-| A `Map` that remembers the original ordering of keys
+
+    This is primarily used so that formatting preserves field order
+
+    This is done primarily to avoid a dependency on @insert-ordered-containers@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
```

Comment:
could the constructor be exported? it will let users extend it with new ops and automatic class deriving. `dhall-to-cabal` would need to add some operators and instances for `Semigroup` and `Monoid`

Reply:
I would like to avoid exporting the constructor so that we can easily change it if other representations end up being more efficient.  I'd prefer to just add whatever `dhall-to-cabal` needs, including the `Semigroup`/`Monoid` instance if necessary

Reply:
Nice! I write down the list of operations used by dhall-to-cabal [here](https://github.com/dhall-lang/dhall-haskell/pull/612#issuecomment-426176497)